File: devops-scm-api/src/main/java/com/tencent/devops/scm/api/pojo/repository/git/GitRepositoryUrl.java
package com.tencent.devops.scm.api.pojo.repository.git;

import java.util.regex.Matcher;
import java.util.regex.Pattern;
import lombok.Data;

@Data
public class GitRepositoryUrl {
    public static final String PROTOCOL_HTTP_SEPARATOR = "://";

    public static final String PROTOCOL_GIT_SEPARATOR = "@";

    public static final String PROTOCOL_GIT = "git";

    public static final String PROTOCOL_SSH = "ssh";

    public static final String PROTOCOL_HTTP = "http";

    public static final String PROTOCOL_HTTPS = "https";

    public static final String PROTOCOL_NONE = "";

    private static final Pattern HOST_AND_PORT_EXTRACTOR =
        Pattern.compile("([^:/\\\\~]*)(?::(\\d*))?(?:([:/\\\\~])(.*))?");

    private String url;

    private String protocol;

    private String userName;

    private String password;

    private String host;

    private String port;
    // 代码库组
    private String group;
    // 代码库名
    private String name;
    // 代码库全名,组+名
    private String fullName;

    public GitRepositoryUrl(String url) {
        this.url = url;

        url = parseProtocol(url);
        url = parseUserInfo(url);
        url = parseHostAndPort(url);
        parseGroupAndName(url);
    }

    private String parseProtocol(String url) {
        if (url.startsWith(PROTOCOL_HTTPS + PROTOCOL_HTTP_SEPARATOR)) {
            this.protocol = PROTOCOL_HTTPS;
            url = url.substring(protocol.length() + 3);
        } else if (url.startsWith(PROTOCOL_HTTP + PROTOCOL_HTTP_SEPARATOR)) {
            this.protocol = PROTOCOL_HTTP;
            url = url.substring(protocol.length() + 3);
        } else if (url.startsWith(PROTOCOL_SSH + PROTOCOL_GIT_SEPARATOR)) {
            this.protocol = PROTOCOL_SSH;
            url = url.substring(protocol.length() + 1);
        } else if (url.startsWith(PROTOCOL_GIT + PROTOCOL_GIT_SEPARATOR)) {
            this.protocol = PROTOCOL_GIT;
            url = url.substring(protocol.length() + 1);
        } else {
            this.protocol = PROTOCOL_NONE;
            return url;
        }

        return url;
    }

    private String parseUserInfo(String url) {
        int indexAt = url.lastIndexOf('@');
        if (indexAt >= 0) {
            String userInfo = url.substring(0, indexAt);
            int indexPwdSep = userInfo.indexOf(':');
            if (indexPwdSep < 0) {
                this.userName = userInfo;
            } else {
                this.userName = userInfo.substring(0, indexPwdSep);
                this.password = userInfo.substring(indexPwdSep + 1);
            }

            url = url.substring(indexAt + 1);
        }
        return url;
    }

    private String parseHostAndPort(String url) {

        Matcher hostAndPortMatcher = HOST_AND_PORT_EXTRACTOR.matcher(url);
        if (hostAndPortMatcher.matches()) {
            if (hostAndPortMatcher.groupCount() > 1 && hostAndPortMatcher.group(1) != null) {
                this.host = hostAndPortMatcher.group(1);
            }
            if (hostAndPortMatcher.groupCount() > 2 && hostAndPortMatcher.group(2) != null) {
                this.port = hostAndPortMatcher.group(2);
            }

            StringBuilder computedUrl = new StringBuilder();
            if (hostAndPortMatcher.group(hostAndPortMatcher.groupCount()) != null) {
                computedUrl.append(hostAndPortMatcher.group(hostAndPortMatcher.groupCount()));
            }
            return computedUrl.toString();
        }
        return url;
    }

    private void parseGroupAndName(String url) {
        int indexAt = url.lastIndexOf(".git");
        if (indexAt >= 0) {
            this.fullName = url.substring(0, indexAt);
        } else {
            this.fullName = url;
        }
        int indexNameAt = fullName.lastIndexOf("/");
        this.group = fullName.substring(0, indexNameAt);
        this.name = fullName.substring(indexNameAt + 1);
    }
}

File: devops-scm-api/src/main/kotlin/com/tencent/devops/scm/api/pojo/repository/git/GitRepositoryUrl.kt
package com.tencent.devops.scm.api.pojo.repository.git

import java.util.regex.Pattern

/**
 * Git仓库URL解析类
 */
data class GitRepositoryUrl(val url: String) {
    companion object {
        const val PROTOCOL_HTTP_SEPARATOR = "://"
        const val PROTOCOL_GIT_SEPARATOR = "@"
        const val PROTOCOL_GIT = "git"
        const val PROTOCOL_SSH = "ssh"
        const val PROTOCOL_HTTP = "http"
        const val PROTOCOL_HTTPS = "https"
        const val PROTOCOL_NONE = ""

        private val HOST_AND_PORT_EXTRACTOR =
            Pattern.compile("([^:/\\\\~]*)(?::(\\d*))?(?:([:/\\\\~])(.*))?")
    }

    val protocol: String
    val userName: String?
    val password: String?
    val host: String
    val port: String?
    val group: String
    val name: String
    val fullName: String

    // 解析过程中的url
    private var processingUrl = url

    init {
        protocol = parseProtocol()
        val userInfo = parseUserInfo()
        userName = userInfo.first
        password = userInfo.second

        val hostInfo = parseHostAndPort()
        host = hostInfo.first
        port = hostInfo.second

        val nameInfo = parseNameInfo()
        fullName = nameInfo.first
        group = nameInfo.second
        name = nameInfo.third
    }

    private fun parseProtocol(): String {
        return when {
            url.startsWith(PROTOCOL_HTTPS + PROTOCOL_HTTP_SEPARATOR) -> {
                processingUrl = url.substring(PROTOCOL_HTTPS.length + PROTOCOL_HTTP_SEPARATOR.length)
                PROTOCOL_HTTPS
            }

            url.startsWith(PROTOCOL_HTTP + PROTOCOL_HTTP_SEPARATOR) -> {
                processingUrl = url.substring(PROTOCOL_HTTP.length + PROTOCOL_HTTP_SEPARATOR.length)
                PROTOCOL_HTTP
            }

            url.startsWith(PROTOCOL_SSH + PROTOCOL_GIT_SEPARATOR) -> {
                processingUrl = url.substring(PROTOCOL_SSH.length + PROTOCOL_GIT_SEPARATOR.length)
                PROTOCOL_SSH
            }

            url.startsWith(PROTOCOL_GIT + PROTOCOL_GIT_SEPARATOR) -> {
                processingUrl = url.substring(PROTOCOL_GIT.length + PROTOCOL_GIT_SEPARATOR.length)
                PROTOCOL_GIT
            }

            else -> {
                PROTOCOL_NONE
            }
        }
    }

    private fun parseUserInfo(): Pair<String?, String?> {
        val indexAt = processingUrl.lastIndexOf('@')
        if (indexAt >= 0) {
            val userInfo = processingUrl.substring(0, indexAt)
            val indexPwdSep = userInfo.indexOf(':')
            processingUrl = url.substring(indexAt + 1)
            if (indexPwdSep < 0) {
                Pair(userInfo, null)
            } else {
                val userName = userInfo.substring(0, indexPwdSep)
                val password = userInfo.substring(indexPwdSep + 1)
                Pair(userName, password)
            }
        }
        return Pair(null, null)
    }

    private fun parseHostAndPort(): Pair<String, String?> {
        val hostAndPortMatcher = HOST_AND_PORT_EXTRACTOR.matcher(processingUrl)
        if (hostAndPortMatcher.matches()) {
            val host = hostAndPortMatcher.group(1)
            val port = if (hostAndPortMatcher.groupCount() > 2 && hostAndPortMatcher.group(2) != null) {
                hostAndPortMatcher.group(2)
            } else {
                null
            }

            val computedUrl = StringBuilder()
            if (hostAndPortMatcher.group(hostAndPortMatcher.groupCount()) != null) {
                computedUrl.append(hostAndPortMatcher.group(hostAndPortMatcher.groupCount()))
            }
            processingUrl = computedUrl.toString()
            return Pair(host, port)
        } else {
            throw IllegalArgumentException("invalid url:$url")
        }
    }

    private fun parseNameInfo(): Triple<String, String, String> {
        val fullName = if (processingUrl.lastIndexOf(".git") >= 0) {
            processingUrl.substring(0, processingUrl.lastIndexOf(".git"))
        } else {
            processingUrl
        }

        val indexNameAt = fullName.lastIndexOf("/")
        val group = fullName.substring(0, indexNameAt)
        val name = fullName.substring(indexNameAt + 1)
        return Triple(fullName, group, name)
    }
}
